File: Xholon/src/org/primordion/user/app/climatechange/carboncycle03/Downwelling.java
package org.primordion.user.app.climatechange.carboncycle03;

import org.primordion.xholon.base.IXholon;

public class Downwelling extends Xhcarboncycle03 {
	
	private double rateFactor = 0.0;
	
	private IXholon surfaceOcean = null;
	private IXholon deepOcean = null;
	
	public void postConfigure() {
		surfaceOcean = ((Flows)getParentNode()).getSurfaceOcean();
		deepOcean = ((Flows)getParentNode()).getDeepOcean();
		rateFactor = this.getVal() / surfaceOcean.getVal();
		super.postConfigure();
	}
	
	public void act() {
		double rate = this.getVal();
		surfaceOcean.decVal(rate);
		deepOcean.incVal(rate);
		super.act();
	}
	
	public void postAct() {
		this.setVal(surfaceOcean.getVal() * rateFactor);
		super.postAct();
	}

}

File: Xholon/war/config/climatechange/carboncycle03/script/javascript/Downwelling.js
<Downwelling val_double="90.2" implName="org.primordion.user.app.climatechange.carboncycle03.Flow_javascript"><![CDATA[
  var behaviorObject = new Object();
  
  behaviorObject.rateFactor = 0.0;
  behaviorObject.surfaceOcean = null;
  behaviorObject.deepOcean = null;
  
  behaviorObject.postConfigure = function() {
    surfaceOcean = contextNodeKey.getParentNode().getSurfaceOcean();
    deepOcean = contextNodeKey.getParentNode().getDeepOcean();
    rateFactor = contextNodeKey.getVal() / surfaceOcean.getVal();
  }
  
  behaviorObject.act = function() {
    var rate = contextNodeKey.getVal();
    surfaceOcean.decVal(rate);
    deepOcean.incVal(rate);
  }
  
  behaviorObject.postAct = function() {
    contextNodeKey.setVal(surfaceOcean.getVal() * rateFactor);
  }

]]></Downwelling>